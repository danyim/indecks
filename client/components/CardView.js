import React from 'react';
import Card from './Card';
import DeckNavigator from './DeckNavigator';

const CardView = React.createClass({
  render() {
    const { deckId, cardIndex } = this.props.params;
    // index of the deck
    const i = this.props.decks.findIndex((deck) => deck.id === deckId)
    // get us the post
    const deck = this.props.decks[i];
    const mode = 'view';

    return (
      <section className="single">
        <div className="single-card">
          <Card card={deck.cards[cardIndex - 1]} />
        </div>

        <DeckNavigator deck={deck} cardIndex={parseInt(cardIndex)} mode={mode} />
      </section>
    )
  }
});

export default CardView;
