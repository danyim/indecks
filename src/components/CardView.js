import React from 'react';
import Swipeable from 'react-swipeable';
import { browserHistory } from 'react-router';
import Card from './Card';
import DeckNavigator from './DeckNavigator';
import styles from '../styles/components/CardView';

class CardView extends React.Component {
  constructor (props) {
    super(props);
    this.state = { flipped: false };

    this.handleEditCard = this.handleEditCard.bind(this);
    this.handleNextCard = this.handleNextCard.bind(this);
    this.handlePrevCard = this.handlePrevCard.bind(this);
    this.handleFlip = this.handleFlip.bind(this);
    this.handleShuffleToggle = this.handleShuffleToggle.bind(this);
  }

  randomCardIndex() {
    return Math.floor((Math.random() * this.maxCardIndex) + 1);
  }

  handleNextCard() {
    if(this.props.cardIndex < this.props.deck.cards.length || this.props.config.shuffle === true) {
      this.handleFlip(null, false);
      let nextIndex = this.props.cardIndex + 1; // Going forwards
      if(this.props.config.shuffle === true) {
        nextIndex = this.randomCardIndex();
      }

      browserHistory.push(`/view/${this.props.deck.id}/${nextIndex}`);
    }
  }

  handlePrevCard() {
    if(this.props.cardIndex > 1 || this.props.config.shuffle === true) {
      this.handleFlip(null, false);
      let nextIndex = this.props.cardIndex - 1; // Going backwards
      if(this.props.config.shuffle === true) {
        nextIndex = this.randomCardIndex();
      }
      browserHistory.push(`/view/${this.props.deck.id}/${nextIndex}`);
    }
  }

  handleEditCard() {
    browserHistory.push(`/edit/${this.props.deck.id}/${this.props.cardIndex}`);
  }

  handleFlip(e, flipped) {
    if(flipped !== undefined) {
      this.setState({ flipped: flipped });
    }
    else {
      this.setState({ flipped: !this.state.flipped });
    }
  }

  handleShuffleToggle() {
    this.props.toggleShuffle();
  }

  render() {
    const { card, deck, config, cardIndex } = this.props;
    const mode = 'view';

    return (
      <section className="single">
        <Swipeable
          onSwipedUp={() => this.handleFlip()}
          onSwipedDown={this.handleEditCard}
          onSwipedLeft={this.handleNextCard}
          onSwipedRight={this.handlePrevCard}>
          <Card card={card} flipped={this.state.flipped} handleOnClick={this.handleFlip} />
        </Swipeable>
        <DeckNavigator deck={deck} cardIndex={cardIndex}
          mode={mode}
          flipped={this.state.flipped}
          config={config}
          handleFlip={this.handleFlip}
          handleShuffleToggle={this.handleShuffleToggle} />
      </section>
    )
  }
}

export default CardView;
