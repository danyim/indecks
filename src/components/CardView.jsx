import React from 'react'
import PropTypes from 'prop-types'
import Swipeable from 'react-swipeable'
import { CardShape, DeckShape } from './__commonShapes'
import Card from './Card'
import DeckNavigator from './DeckNavigator'
// import styles from '../styles/components/CardView.styl'

class CardView extends React.Component {
  static propTypes = {
    card: CardShape.isRequired,
    deck: DeckShape.isRequired,
    config: PropTypes.shape({ shuffle: PropTypes.bool.isRequired }).isRequired,
    cardIndex: PropTypes.number.isRequired,
    shuffleOn: PropTypes.func.isRequired,
    shuffleOff: PropTypes.func.isRequired,
    push: PropTypes.func.isRequired
  }

  static defaultProps = {}

  static randomCardIndex(max) {
    return Math.floor(Math.random() * max + 1)
  }

  constructor(props) {
    super(props)
    this.state = { flipped: false }

    this.handleEditCard = this.handleEditCard.bind(this)
    this.handleNextCard = this.handleNextCard.bind(this)
    this.handlePrevCard = this.handlePrevCard.bind(this)
    this.handleFlip = this.handleFlip.bind(this)
    this.handleShuffleToggle = this.handleShuffleToggle.bind(this)
  }

  handleNextCard() {
    if (
      this.props.cardIndex < this.props.deck.cards.length ||
      this.props.config.shuffle === true
    ) {
      this.handleFlip(null, false)
      let nextIndex = this.props.cardIndex + 1 // Going forwards
      if (this.props.config.shuffle === true) {
        nextIndex = CardView.randomCardIndex(this.props.deck.cards.length)
      }

      this.props.push(`/view/${this.props.deck.id}/${nextIndex}`)
    }
  }

  handlePrevCard() {
    if (this.props.cardIndex > 1 || this.props.config.shuffle === true) {
      this.handleFlip(null, false)
      let nextIndex = this.props.cardIndex - 1 // Going backwards
      if (this.props.config.shuffle === true) {
        nextIndex = CardView.randomCardIndex(this.props.deck.cards.length)
      }
      this.props.push(`/view/${this.props.deck.id}/${nextIndex}`)
    }
  }

  handleEditCard() {
    this.props.push(`/edit/${this.props.deck.id}/${this.props.cardIndex}`)
  }

  /**
   * Flips the card
   */
  handleFlip(e, flipped = null) {
    if (flipped !== null) {
      this.setState({ flipped })
    } else {
      this.setState({ flipped: !this.state.flipped })
    }
  }

  handleShuffleToggle() {
    if (this.props.config.shuffle === true) {
      this.props.shuffleOff()
    } else {
      this.props.shuffleOn()
    }
  }

  render() {
    const { card, deck, config, cardIndex } = this.props
    const mode = 'view'

    return (
      <section className="single">
        <Swipeable
          // onSwiping={() => console.log('swiping')}
          // onSwiped={() => console.log('swiped')}
          onSwipedLeft={this.handleNextCard}
          onSwipedRight={this.handlePrevCard}
        >
          <Card
            card={card}
            flipped={this.state.flipped}
            handleOnClick={this.handleFlip}
          />
        </Swipeable>
        <DeckNavigator
          deck={deck}
          cardIndex={cardIndex}
          mode={mode}
          push={this.props.push}
          shuffle={config.shuffle}
          flipped={this.state.flipped}
          handleFlip={this.handleFlip}
          handleNextCard={this.handleNextCard}
          handlePrevCard={this.handlePrevCard}
          handleShuffleToggle={this.handleShuffleToggle}
        />
      </section>
    )
  }
}

export default CardView
