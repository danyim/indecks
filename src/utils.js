export const createReducer = (initialState, handlers) => {
  return function reducer(state = initialState, action = {}) {
    if (handlers.hasOwnProperty(action.type)) {
      return handlers[action.type](state, action);
    }
    return state;
  };
};

/**
 * Given an array of consecutive numbers, find the missing value.
 * Returns null if no missing value found
 * @param  {Array}  a Array of numbers
 * @return {Number}   Missing value
 */
export const findMissing = (a) => {
  let expectedNextValue;
  for(let k of a)
  {
    expectedNextValue = a[k] + 1;
    if(k <= a.length - 1 && expectedNextValue !== a[k + 1]) {
      return expectedNextValue;
    }
  }
  return null;
};

/**
 * Given an array with potentially missing values at parameter key, find the
 * next best 0-based index to insert the next value
 * @param  {Array}  a   An array of objects
 * @param  {String} key Key to look for
 * @return {Number}     Index of the best next value
 */
export const findBestNextIndex = (a, key) => {
  return findMissing(a.map(x => x[key])) || Math.max(0, Math.max.apply(null, a.map(x => x[key])) + 1);
};

