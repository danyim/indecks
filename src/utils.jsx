import React from 'react'
import { Route } from 'react-router-dom'
import _ from 'lodash'

export const createReducer = (initialState, handlers) =>
  function reducer(state = initialState, action = {}) {
    if (Object.prototype.hasOwnProperty.call(handlers, action.type)) {
      return handlers[action.type](state, action)
    }
    return state
  }

/**
 * Given an array of consecutive numbers from 1 to n, find the missing value.
 * Returns null if no missing value found
 * @param  {Array}  a   Array of numbers
 * @param  {Number} min Minimum value (0 or 1)
 * @return {Number}     Missing value
 */
export const findMissing = (a, min = 1) => {
  let expected = min
  for (let k = 0; k < a.length; k++) {
    if (a[k] !== expected) {
      return expected
    }
    expected++
  }
  return null
}

/**
 * Given an array of objects, check the value of the key and return the
 * best 1-based index to insert the next value
 * @param  {Array}  a   An array of objects
 * @param  {String} key Key to look for
 * @return {Number}     Index of the best next value
 */
export const findBestNextIndex = (a, key) =>
  findMissing(a.map(x => x[key])) ||
  Math.max(0, Math.max.apply(null, a.map(x => x[key])) + 1)

/**
 * Helper function to generate routes
 * @param  {Object} route) Route object that takes the form
 *                         {
 *                           exact: boolean,
 *                           path: string,
 *                           component: Component,
 *                           routes: Routes
 *                         }
 * @return {Component}        Returns a <Route> component to plug into a <Switch>
 */
export const routeWithSubRoutes = route =>
  <Route
    key={_.uniqueId()}
    exact={route.exact || false}
    path={route.path}
    render={props =>
      // Pass the sub-routes down to keep nesting
      // <route.component {...props}>
      //   { route.children ? route.children.map(child => routeWithSubRoutes(child)) : null }
      // </route.component>
      <route.component {...props} routes={route.routes} />}
  />
