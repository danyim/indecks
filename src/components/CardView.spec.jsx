import toJSON from 'enzyme-to-json'
import { shallow, mount, setup, setupFull } from '../test/utils' // eslint-disable-line no-unused-vars
import { createStartTouchEventObject, createMoveTouchEventObject } from '../test/eventHelpers.js'
import CardView from './CardView'

describe('CardView', () => {
  let props

  beforeEach(() => {
    props = {
      card: {
        index: 1,
        title: 'Card',
        answer: 'Card description',
      },
      deck: {
        id: 'ABCDEFG',
        title: 'Deck title',
        description: 'Deck description',
        cards: [
          {
            index: 1,
            title: 'Card',
            answer: 'Card description',
          },
        ],
      },
      config: {
        shuffle: false,
      },
      cardIndex: 1,
      shuffleOff: () => {},
      shuffleOn: () => {},
      push: () => {},
    }
  })
  it('should render self and subcomponents', () => {
    const tree = shallow(CardView, props)
    expect(toJSON(tree)).toMatchSnapshot()
  })

  it('should display a card', () => {
    const { wrapper } = setupFull(CardView, props)
    const subject = wrapper.find('Card')
    expect(subject.exists()).toBe(true)
  })

  it('should flip the card when clicked', () => {
    const { wrapper } = setupFull(CardView, props)
    wrapper.find('Card').simulate('click')
    expect(wrapper.state('flipped')).toBe(true)
  })

  // TODO: Capturing swipe events still needs some work
  xit('should advance to the next card when swiped left', () => {
    const handler = jest.fn()
    const { wrapper } = setupFull(CardView, {
      ...props,
      push: handler,
    })
    const subject = wrapper.find('Swipeable')
    expect(subject.exists()).toBe(true)
    subject.simulate('touchStart', createStartTouchEventObject({ x: 200, y: 0 }))
    subject.simulate('touchMove', createMoveTouchEventObject({ x: 150, y: 0 }))
    subject.simulate('touchEnd', createMoveTouchEventObject({ x: 100, y: 0 }))
    expect(handler).toHaveBeenCalled()
  })

  xit('should go to the previous card when swiped right', () => {
    const handler = jest.fn()
    const { wrapper } = setupFull(CardView, {
      ...props,
      push: handler,
    })
    const subject = wrapper.find('Swipeable')
    expect(subject.exists()).toBe(true)
    subject.simulate('touchStart', createStartTouchEventObject({ x: 100, y: 0 }))
    subject.simulate('touchMove', createMoveTouchEventObject({ x: 150, y: 0 }))
    subject.simulate('touchEnd', createMoveTouchEventObject({ x: 200, y: 0 }))
    expect(handler).toHaveBeenCalled()
  })
})
