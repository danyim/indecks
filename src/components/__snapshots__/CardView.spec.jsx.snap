// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardView should render self and subcomponents 1`] = `
<section
  className="single"
>
  <div
    onMouseDown={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
  >
    <figure
      className="grid-figure grid-figure "
      onClick={[Function]}
      role="presentation"
    >
      <div
        className="card-contents card-contents"
      >
        <div
          className="card-title card-title"
          dangerouslySetInnerHTML={
            Object {
              "__html": "<p>Card</p>
            ",
            }
          }
          onClick={null}
        />
      </div>
    </figure>
  </div>
  <div
    className="deck-navigator"
  >
    <div
      className="left"
    >
      <div
        className="frontBack"
      >
        <a
          className="front active"
          onClick={[Function]}
          role="presentation"
          title="View front of card"
        >
          FRONT
        </a>
        <span
          className="separator"
        >
          |
        </span>
        <a
          className="back"
          onClick={[Function]}
          role="presentation"
          title="View back of card"
        >
          BACK
        </a>
      </div>
    </div>
    <div
      className="deck-nav-controls"
    >
      <button
        className="button"
        disabled={true}
        onClick={[Function]}
        title="Previous Card"
      >
        <i
          className="fa fa-backward"
        />
      </button>
      <button
        className="button btn-inactive"
        onClick={[Function]}
        title="Shuffle"
      >
        <i
          className="fa fa-random"
        />
      </button>
      <button
        className="button"
        disabled={true}
        onClick={[Function]}
        title="Next Card"
      >
        <i
          className="fa fa-forward"
        />
      </button>
    </div>
    <div
      className="right"
    >
      <div
        className="cardCount"
      >
        <span
          className="current"
        >
          1
        </span>
        <span
          className="separator"
        />
        <span
          className="max"
        >
          1
        </span>
      </div>
    </div>
  </div>
</section>
`;
